### ISSUE TITLE
Inquiry Regarding Dataset Construction

### ISSUE THREAD

**Issue by L1-M1ng** [siavashzk/BranchNet#11]
I have some questions about the dataset construction in our code repository, and I would greatly appreciate your insights. Here are my inquiries:

**1. Current Data Set Construction:**
As of now, my dataset construction process is as follows:
I have used the leela_s program from spec2017 as an example. Initially, I executed the program with the three default inputs provided by spec2017, namely train, ref, and test, resulting in three traces.
Subsequently, I proceeded to partition the dataset.
![image](https://github.com/siavashzk/BranchNet/assets/54474696/a02662ba-6314-48c1-8c3f-98047238aa01)
However, as per the original research paper's recommendation, I attempted to select "hard-to-predict branches" from the validation set and encountered situations where these branches were not retrievable from the training set. Could you please offer guidance on how to address this issue?
**2. Regarding Alberta Inputs:** 
I've reviewed the scenario where Alberta inputs are used for training. Alberta provides a wide range of inputs. Do I need to generate traces for all of these inputs to use them as part of the training set?
**3.About the weight Parameter in benchmark.yaml:**
I've noticed that the benchmark.yaml file includes a weight parameter within the simpoint section. I am curious about the significance of this parameter and how one should obtain it.
Thank you in advance for your assistance. Your expertise will be invaluable in resolving these dataset construction queries.

**Comment by siavashzk** [comment]
Sorry for the late reply.

1: do you mean that the branches that were deemed hard-to-predict in your validation set were not available in your training set? It seems like you need to expand your inputs to the programs to have higher coverage to generate better datasets.

2: The more inputs you have the better. The paper shows that the training accuracy improves with higher coverage in the training set. (although there is a point of diminishing return). For example, for the benchmark leela, training on 2 or 3 of Alberta inputs was enough to hit close to the same accuracy as training on all of Alberta inputs. (See Figure 12 in the paper)

3: The weights would be provided by SimPoints. You can read about what those weights mean by reading the simpoints paper:
https://cseweb.ucsd.edu/~calder/papers/ASPLOS-02-SimPoint.pdf

If your methodology for getting sample traces for your simulation does not involve weights (e.g., if you're using random time-based sampling), then you can simply assign equal weights to all your traces.

**Comment by L1-M1ng** [comment]
You're welcome! I appreciate your reply very much!

For Q1: I have already resolved this issue through static compilation.
For Q2: I am trying to incorporate Alberta inputs to generate sufficient data for training. I used train input in SPEC2017 leela_r program, but just reach 92% accuracy with Big-Branch. Is this result normal?
For Q3:  I have learned about SimPoint, it requires a  .bb (Basic Block ) file for clustering. How did you obtain this file? It seems there are no scripts related to this in the repository. 
Q4: After obtaining SimPoint weights, how should Pin Tool be used for tracing?

I'm sorry, I have quite a few questions. In the future, maybe when my research results are written into a paper and published, I will mention you in the acknowledgments. . Once again, I sincerely appreciate your answers.

**Comment by L1-M1ng** [comment]
My e-mail is m.li@i4ai.org

**Comment by siavashzk** [comment]
Yes, static linking and turning off ASLR is needed to get stable PCs for offline training. In theory, the same could be achieving with additional metadata attached to the binaries to support dynamic linking & ASLR but I haven't done any work on that.

Q2: I don't remember the exact number. But from what I recall, the combination of 64KB TAGE-SC-L + Big-BranchNet for leela_r was higher. The MPKI reduction numbers are in the paper.

Q3: Yes, I do not have the copyright to redistribute the simpoint scripts so I left them out of right repo. There are many other sources for generating simpoints online if you search.  What I did specially was to use Intel pinplay
https://www.intel.com/content/www/us/en/developer/articles/tool/program-recordreplay-toolkit.html
It comes (or came) with a wrapper script for getting pinballs for simpoints for the programs. Then I had scripts (available in this repo), that ran those pinballs and produced branch traces.

Q4: refer to pintool tutorials online. I don't have time to support tools and scripts but I'm happy to answer questions about the research.

**Comment by L1-M1ng** [comment]
Thanks to your guidance, I have successfully used simpoint to get the results of the cluster and collect the data set.
I replicated your work on leela_r benchmark, but the result of tage_sc_l_64 obtained by using run_tage.py is very accurate, and the result of top10 hard branch on ref workload is as follows:
![image](https://github.com/siavashzk/BranchNet/assets/54474696/e9a536b7-6756-4b42-a554-c7a1e23cb8f4)
(This result seem unusually high? In your paper, tage_sc_l_64 is probably 70%~80% accurate for many hard branches in leela_s)
![image](https://github.com/siavashzk/BranchNet/assets/54474696/967e70a4-dbc7-42c3-a812-88758366a7f5)
Then I used ref workload for training and testing, and selected the Big-BranchNet model to see whether the model could be overfitted, but it seemed that the model only reached about 90% accuracy, and the prediction result of hard branch top1 was as follows.
![image](https://github.com/siavashzk/BranchNet/assets/54474696/8ff0fe9b-b7a1-4041-b68a-d71fa0c533ff)
The training setting:
![image](https://github.com/siavashzk/BranchNet/assets/54474696/11b12e15-b4bd-474e-93bf-bfa9a5d18887)
Could you tell me where I might have gone wrong? (I am collecting the data set of leela_s)

Best wishes!

**Comment by L1-M1ng** [comment]
![image](https://github.com/siavashzk/BranchNet/assets/54474696/6cc37480-c030-4bd1-aa85-6cdb1cebc0a0)
I got the result of tage_sc_l_64 on leela_s of ref workload, it seems the same as the result on leela_r

**Comment by Jurisu25** [comment]
> ![图像](https://private-user-images.githubusercontent.com/54474696/293204179-6cc37480-c030-4bd1-aa85-6cdb1cebc0a0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY5NDg3NzgsIm5iZiI6MTczNjk0ODQ3OCwicGF0aCI6Ii81NDQ3NDY5Ni8yOTMyMDQxNzktNmNjMzc0ODAtYzAzMC00YmQxLWFhODUtNmNkYjFjZWJjMGEwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTE1VDEzNDExOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI0OTFiOWI0YzE5YjBkN2QzZDRhMWRlOGRlMjcwN2RiOTU0MDY2OWZhYjMyMGE1YTU4Yjg3ZGE4MDU3ZDE1NTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.iYHOCiYsxywBeKYG_8_k1a2Ewb074e9AEy0u-RTCXhQ) 我在 ref 工作负载的 leela_s 上得到了 tage_sc_l_64 的结果，它似乎与 leela_r 上的结果相同

Sorry to bother you, I am a bit confused about the part of generating datasets from Simpoint. Can you briefly guide me on how to generate datasets for spec2017

**Comment by Cipher-M-0** [comment]
> ![image](https://github.com/siavashzk/BranchNet/assets/54474696/6cc37480-c030-4bd1-aa85-6cdb1cebc0a0) I got the result of tage_sc_l_64 on leela_s of ref workload, it seems the same as the result on leela_r

Hello, may I ask if you were eventually able to run it successfully and obtain reasonably good results?
